Test instability On Windows builds

https://ci.appveyor.com/project/AndrewAnnex/spiceypy/build/1.1.707

looks like something strange is occurring with windows builds. 

```
          epochs = stypes.toDoubleVector(epochs)
>       libspice.spk14a_c(handle, ncsets, coeffs, epochs)
E       WindowsError: exception: access violation reading 0x000000000367F000
spiceypy\spiceypy.py:11151: WindowsError
_________________________________ test_spk14b _________________________________

```

--- a/.travis.yml
+++ b/.travis.yml
@@ -43,7 +43,6 @@ before_install:
 - pip install --use-wheel -r requirements.txt
 install:
 - python setup.py install
-- pytest -s
 script:
 - coverage run --source=spiceypy setup.py test
 after_success:

--- a/.travis.yml
+++ b/.travis.yml
@@ -43,6 +43,7 @@ before_install:
 - pip install --use-wheel -r requirements.txt
 install:
 - python setup.py install
+- pytest -s
 script:
 - coverage run --source=spiceypy setup.py test
 after_success:

--- a/.travis.yml
+++ b/.travis.yml
@@ -19,10 +19,10 @@ matrix:
           env: PYVERSION='2.7.13'
         - os: osx
           language: generic
-          env: PYVERSION='3.4.0'
+          env: PYVERSION='3.4.7'
         - os: osx
           language: generic
-          env: PYVERSION='3.5.0'
+          env: PYVERSION='3.5.4'
         - os: osx
           language: generic
           env: PYVERSION='3.6.0'

--- a/.travis.yml
+++ b/.travis.yml
@@ -44,7 +44,7 @@ before_install:
 install:
 - python setup.py install
 script:
-- if [ "$TRAVIS_OS_NAME" == "osx" ]; then travis_wait 30 coverage run --source=spiceypy setup.py test; else coverage run --source=spiceypy setup.py test; fi
+- coverage run --source=spiceypy setup.py test
 after_success:
 - coveralls
 deploy:

--- a/spiceypy/spiceypy.py
+++ b/spiceypy/spiceypy.py
@@ -3765,14 +3765,8 @@ def ekbseg(handle, tabnam, cnames, decls):
     :type handle: int
     :param tabnam: Table name.
     :type tabnam: str
-    :param ncols: Number of columns in the segment.
-    :type ncols: int
-    :param cnmlen: Length of names in in column name array.
-    :type cnmlen: int
     :param cnames: Names of columns.
     :type cnames: list of str.
-    :param declen: Length of declaration strings in declaration array.
-    :type declen: int
     :param decls: Declarations of columns.
     :type decls: list of str.
     :return: Segment number.

--- a/spiceypy/spiceypy.py
+++ b/spiceypy/spiceypy.py
@@ -3517,7 +3517,7 @@ def edterm(trmtyp, source, target, et, fixref, abcorr, obsrvr, npts):
 
 
 @spiceErrorCheck
-def ekacec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
+def ekacec(handle, segno, recno, column, nvals, cvals, isnull):
     """
     Add data to a character column in a specified EK record.
 
@@ -3533,8 +3533,6 @@ def ekacec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
     :type column: str
     :param nvals: Number of values to add to column.
     :type nvals: int
-    :param vallen: Declared length of character values.
-    :type vallen: int
     :param cvals: Character values to add to column.
     :type cvals: list of str.
     :param isnull: Flag indicating whether column entry is null.
@@ -3545,11 +3543,10 @@ def ekacec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
     recno = ctypes.c_int(recno)
     column = stypes.stringToCharP(column)
     nvals = ctypes.c_int(nvals)
-    vallen = ctypes.c_int(vallen)
-    cvals = stypes.listToCharArrayPtr(cvals,xLen=vallen.value)
+    vallen = ctypes.c_int(len(max(cvals, key=len)) + 1)
+    cvals = stypes.listToCharArrayPtr(cvals)
     isnull = ctypes.c_int(1 if isnull else 0)
-    libspice.ekacec_c(handle, segno, recno, column, nvals, vallen, cvals,
-                      isnull)
+    libspice.ekacec_c(handle, segno, recno, column, nvals, vallen, cvals, isnull)
 
 
 @spiceErrorCheck
@@ -3758,19 +3755,7 @@ def ekappr(handle, segno):
 
 
 @spiceErrorCheck
-def ekbseg(handle, tabnam, ncols, cnmlen, cnames, declen, decls):
-    # if 'cnmlen' in kwargs:
-    # cnmlen = kwargs['cnmlen']
-    # else:
-    # cnmlen = len(max(cnames, key=len)) + 1
-    # if 'declen' in kwargs:
-    #     declen = kwargs['declen']
-    # else:
-    #     declen = len(max(decls, key=len)) + 1
-    # if 'ncols' in kwargs:
-    #     ncols = kwargs['ncols']
-    # else:
-    #     ncols = len(cnames)
+def ekbseg(handle, tabnam, cnames, decls):
     """
     Start a new segment in an E-kernel.
 
@@ -3795,13 +3780,13 @@ def ekbseg(handle, tabnam, ncols, cnmlen, cnames, declen, decls):
     """
     handle = ctypes.c_int(handle)
     tabnam = stypes.stringToCharP(tabnam)
-    cnmlen = ctypes.c_int(cnmlen)
-    cnames = stypes.listToCharArray(cnames,xLen=cnmlen.value)  # not sure if this works
-    declen = ctypes.c_int(declen)
-    decls = stypes.listToCharArray(decls,xLen=declen.value)
+    ncols  = ctypes.c_int(len(cnames))
+    cnmlen = ctypes.c_int(len(max(cnames, key=len)) + 1) # needs to be len(name)+1 ie 'c1' to 3 for ekbseg do not fail
+    cnames = stypes.listToCharArrayPtr(cnames)
+    declen = ctypes.c_int(len(max(decls, key=len)) + 1)
+    decls = stypes.listToCharArrayPtr(decls)
     segno = ctypes.c_int()
-    libspice.ekbseg_c(handle, tabnam, ncols, cnmlen, cnames, declen, decls,
-                      ctypes.byref(segno))
+    libspice.ekbseg_c(handle, tabnam, ncols, cnmlen, cnames, declen, decls, ctypes.byref(segno))
     return segno.value
 
 
@@ -3934,7 +3919,6 @@ def ekfind(query, lenout=_default_len_out):
 @spiceErrorCheck
 @spiceFoundExceptionThrower
 def ekgc(selidx, row, element, lenout=_default_len_out):
-    # ekgc has issues grabbing last element/row in column
     """
     Return an element of an entry in a column of character type in a specified
     row.
@@ -3960,9 +3944,8 @@ def ekgc(selidx, row, element, lenout=_default_len_out):
     lenout = ctypes.c_int(lenout)
     null = ctypes.c_bool()
     found = ctypes.c_bool()
-    cdata = stypes.stringToCharP(" " * lenout.value)
-    libspice.ekgc_c(selidx, row, element, lenout, cdata, ctypes.byref(null),
-                    ctypes.byref(found))
+    cdata = stypes.stringToCharP(lenout)
+    libspice.ekgc_c(selidx, row, element, lenout, cdata, ctypes.byref(null), ctypes.byref(found))
     return stypes.toPythonString(cdata), null.value, found.value
 
 
@@ -4104,7 +4087,7 @@ def eklef(fname):
     """
     fname = stypes.stringToCharP(fname)
     handle = ctypes.c_int()
-    libspice.eklef_c(fname, handle)
+    libspice.eklef_c(fname, ctypes.byref(handle))
     return handle.value
 
 
@@ -4179,7 +4162,7 @@ def ekopn(fname, ifname, ncomch):
     ifname = stypes.stringToCharP(ifname)
     ncomch = ctypes.c_int(ncomch)
     handle = ctypes.c_int()
-    libspice.ekopn_c(fname, ifname, ncomch, handle)
+    libspice.ekopn_c(fname, ifname, ncomch, ctypes.byref(handle))
     return handle.value
 
 
@@ -4272,15 +4255,13 @@ def ekpsel(query, msglen, tablen, collen):
     xends = stypes.emptyIntVector(_SPICE_EK_MAXQSEL)
     xtypes = stypes.emptyIntVector(_SPICE_EK_MAXQSEL)
     xclass = stypes.emptyIntVector(_SPICE_EK_MAXQSEL)
-    tabs = stypes.emptyCharArray(yLen=_SPICE_EK_MAXQSEL, xLen=tablen.value)
-    cols = stypes.emptyCharArray(yLen=_SPICE_EK_MAXQSEL, xLen=collen.value)
+    tabs = stypes.emptyCharArray(yLen=_SPICE_EK_MAXQSEL, xLen=tablen)
+    cols = stypes.emptyCharArray(yLen=_SPICE_EK_MAXQSEL, xLen=collen)
     error = ctypes.c_bool()
-    errmsg = stypes.stringToCharP(" " * msglen.value)
+    errmsg = stypes.stringToCharP(msglen)
     libspice.ekpsel_c(query, msglen, tablen, collen, ctypes.byref(n),
-                      xbegs, xends,
-                      xtypes, xclass,
-                      ctypes.byref(tabs), ctypes.byref(cols),
-                      ctypes.byref(error), errmsg)
+                      xbegs, xends, xtypes, xclass, ctypes.byref(tabs),
+                      ctypes.byref(cols), ctypes.byref(error), errmsg)
     return (n.value,
             stypes.cVectorToPython(xbegs)[:n.value],
             stypes.cVectorToPython(xends)[:n.value],
@@ -4323,10 +4304,9 @@ def ekrcec(handle, segno, recno, column, lenout, nelts=_SPICE_EK_EKRCEX_ROOM_DEF
     column = stypes.stringToCharP(column)
     lenout = ctypes.c_int(lenout)
     nvals = ctypes.c_int()
-    cvals = stypes.emptyCharArray(yLen=nelts, xLen=lenout.value)
+    cvals = stypes.emptyCharArray(yLen=nelts, xLen=lenout)
     isnull = ctypes.c_bool()
-    libspice.ekrcec_c(handle, segno, recno, column, lenout, ctypes.byref(nvals),
-                      ctypes.byref(cvals), ctypes.byref(isnull))
+    libspice.ekrcec_c(handle, segno, recno, column, lenout, ctypes.byref(nvals), ctypes.byref(cvals), ctypes.byref(isnull))
     assert failed() or (nvals.value <= nelts)
     return nvals.value, stypes.cVectorToPython(cvals)[:nvals.value], isnull.value
 
@@ -4442,7 +4422,7 @@ def ektnam(n, lenout=_default_len_out):
 
 
 @spiceErrorCheck
-def ekucec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
+def ekucec(handle, segno, recno, column, nvals, cvals, isnull):
     """
     Update a character column entry in a specified EK record.
 
@@ -4458,8 +4438,6 @@ def ekucec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
     :type column: str
     :param nvals: Number of values in new column entry.
     :type nvals: int
-    :param vallen: Declared length of character values.
-    :type vallen: int
     :param cvals: Character values comprising new column entry.
     :type cvals: list of str.
     :param isnull: Flag indicating whether column entry is null.
@@ -4470,11 +4448,10 @@ def ekucec(handle, segno, recno, column, nvals, vallen, cvals, isnull):
     recno = ctypes.c_int(recno)
     column = stypes.stringToCharP(column)
     nvals = ctypes.c_int(nvals)
-    vallen = ctypes.c_int(vallen)
-    cvals = stypes.listToCharArrayPtr(cvals,xLen=vallen.value)
+    vallen = ctypes.c_int(len(max(cvals, key=len)) + 1)
+    cvals = stypes.listToCharArrayPtr(cvals, xLen=vallen)
     isnull = ctypes.c_int(1 if isnull else 0)
-    libspice.ekucec_c(handle, segno, recno, column, nvals, vallen, cvals,
-                      isnull)
+    libspice.ekucec_c(handle, segno, recno, column, nvals, vallen, cvals, isnull)
 
 
 @spiceErrorCheck

--- a/spiceypy/tests/test_wrapper.py
+++ b/spiceypy/tests/test_wrapper.py
@@ -1644,6 +1644,7 @@ def test_dskv02():
     spice.kclear()
 
 def test_dskw02_dskrb2_dskmi2():
+    print("in dskw02 test")
     spice.kclear()
     dskpath = os.path.join(cwd, "TESTdskw02.dsk")
     if spice.exists(dskpath):
@@ -1662,12 +1663,12 @@ def test_dskw02_dskrb2_dskmi2():
     first = -50 * spice.jyear()
     last  =  50 * spice.jyear()
     # stuff from spicedsk.h
-    SPICE_DSK02_MAXVRT = 16000002 // 16 # divide to lower memory usage
+    SPICE_DSK02_MAXVRT = 16000002 // 128 # divide to lower memory usage
     SPICE_DSK02_MAXPLT = 2 * (SPICE_DSK02_MAXVRT - 2)
     SPICE_DSK02_MAXVXP = SPICE_DSK02_MAXPLT // 2
-    SPICE_DSK02_MAXCEL = 60000000 // 16 # divide to lower memory usage
+    SPICE_DSK02_MAXCEL = 60000000 // 128 # divide to lower memory usage
     SPICE_DSK02_MXNVLS = SPICE_DSK02_MAXCEL + (SPICE_DSK02_MAXVXP // 2)
-    SPICE_DSK02_MAXCGR = 100000   // 16 # divide to lower memory usage
+    SPICE_DSK02_MAXCGR = 100000   // 128 # divide to lower memory usage
     SPICE_DSK02_IXIFIX = SPICE_DSK02_MAXCGR + 7
     SPICE_DSK02_MAXNPV = 3 * (SPICE_DSK02_MAXPLT // 2) + 1
     SPICE_DSK02_SPAISZ = SPICE_DSK02_IXIFIX + SPICE_DSK02_MAXVXP + SPICE_DSK02_MXNVLS + SPICE_DSK02_MAXVRT + SPICE_DSK02_MAXNPV
@@ -1705,6 +1706,7 @@ def test_dskw02_dskrb2_dskmi2():
     if spice.exists(dskpath):
         os.remove(dskpath)  # pragma: no cover
     spice.kclear()
+    print("done dskw02 test")
 
 
 def test_dskx02():
@@ -2007,6 +2009,7 @@ def test_ekaced():
 
 
 def test_ekmany():
+    print("in ekmany")
     spice.kclear()
     ekpath = os.path.join(cwd, "example_ekmany.ek")
     tablename = "test_table_ekmany"
@@ -2158,6 +2161,7 @@ def test_ekmany():
     assert not spice.failed()
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
+    print("finished ekmany")
 
 
 def test_ekaclc():

--- a/spiceypy/tests/test_wrapper.py
+++ b/spiceypy/tests/test_wrapper.py
@@ -1980,10 +1980,9 @@ def test_ekacec():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "test_table_ekacec", 1, 10, ["c1"], 200,
-                         ["DATATYPE = CHARACTER*(*), NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "test_table_ekacec", ["c1"], ["DATATYPE = CHARACTER*(*), NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
-    spice.ekacec(handle, segno, recno, "c1", 2, 10, ["1.0", "2.0"], False)
+    spice.ekacec(handle, segno, recno, "c1", 2, ["1.0", "2.0"], False)
     spice.ekcls(handle)
     spice.kclear()
     if spice.exists(ekpath):
@@ -1997,8 +1996,7 @@ def test_ekaced():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "test_table_ekaced", 1, 10, ["c1"], 200,
-                         ["DATATYPE = DOUBLE PRECISION, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "test_table_ekaced", ["c1"], ["DATATYPE = DOUBLE PRECISION, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekaced(handle, segno, recno, "c1", 2, [1.0, 2.0], False)
     spice.ekcls(handle)
@@ -2219,8 +2217,7 @@ def test_ekappr():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "test_table_ekappr", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "test_table_ekappr", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)
@@ -2238,7 +2235,7 @@ def test_ekbseg():
     handle = spice.ekopn(ekpath, "Test EK", 100)
     cnames = ['INT_COL_1']
     cdecls = ["DATATYPE=INTEGER, INDEXED=TRUE, NULLS_OK=TRUE"]
-    segno = spice.ekbseg(handle, "SCALAR_DATA", 1, 100, cnames, 200, cdecls)
+    segno = spice.ekbseg(handle, "SCALAR_DATA", cnames, cdecls)
     recno = spice.ekappr(handle, segno)
     assert recno != -1
     ordids = [x for x in range(5)]
@@ -2261,8 +2258,7 @@ def test_ekccnt():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "TEST_TABLE_EKCCNT", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "TEST_TABLE_EKCCNT", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)
@@ -2283,8 +2279,7 @@ def test_ekcii():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "TEST_TABLE_EKCII", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "TEST_TABLE_EKCII", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)
@@ -2498,8 +2493,7 @@ def test_eklef():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "test_table_eklef", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "test_table_eklef", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)
@@ -2517,8 +2511,7 @@ def test_eknseg():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "TEST_TABLE_EKNSEG", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "TEST_TABLE_EKNSEG", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)
@@ -2621,8 +2614,7 @@ def test_ektnam():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, "TEST_TABLE_EKTNAM", 1, 10, ["c1"], 200,
-                         ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
+    segno = spice.ekbseg(handle, "TEST_TABLE_EKTNAM", ["c1"], ["DATATYPE  = INTEGER, NULLS_OK = TRUE"])
     recno = spice.ekappr(handle, segno)
     spice.ekacei(handle, segno, recno, "c1", 2, [1, 2], False)
     spice.ekcls(handle)

--- a/spiceypy/tests/test_wrapper.py
+++ b/spiceypy/tests/test_wrapper.py
@@ -1662,12 +1662,12 @@ def test_dskw02_dskrb2_dskmi2():
     first = -50 * spice.jyear()
     last  =  50 * spice.jyear()
     # stuff from spicedsk.h
-    SPICE_DSK02_MAXVRT = 16000002 // 4 # divide to lower memory usage
+    SPICE_DSK02_MAXVRT = 16000002 // 16 # divide to lower memory usage
     SPICE_DSK02_MAXPLT = 2 * (SPICE_DSK02_MAXVRT - 2)
     SPICE_DSK02_MAXVXP = SPICE_DSK02_MAXPLT // 2
-    SPICE_DSK02_MAXCEL = 60000000 // 4 # divide to lower memory usage
+    SPICE_DSK02_MAXCEL = 60000000 // 16 # divide to lower memory usage
     SPICE_DSK02_MXNVLS = SPICE_DSK02_MAXCEL + (SPICE_DSK02_MAXVXP // 2)
-    SPICE_DSK02_MAXCGR = 100000   // 4 # divide to lower memory usage
+    SPICE_DSK02_MAXCGR = 100000   // 16 # divide to lower memory usage
     SPICE_DSK02_IXIFIX = SPICE_DSK02_MAXCGR + 7
     SPICE_DSK02_MAXNPV = 3 * (SPICE_DSK02_MAXPLT // 2) + 1
     SPICE_DSK02_SPAISZ = SPICE_DSK02_IXIFIX + SPICE_DSK02_MAXVXP + SPICE_DSK02_MXNVLS + SPICE_DSK02_MAXVRT + SPICE_DSK02_MAXNPV

--- a/spiceypy/tests/test_wrapper.py
+++ b/spiceypy/tests/test_wrapper.py
@@ -1644,7 +1644,6 @@ def test_dskv02():
     spice.kclear()
 
 def test_dskw02_dskrb2_dskmi2():
-    print("in dskw02 test")
     spice.kclear()
     dskpath = os.path.join(cwd, "TESTdskw02.dsk")
     if spice.exists(dskpath):
@@ -1706,7 +1705,6 @@ def test_dskw02_dskrb2_dskmi2():
     if spice.exists(dskpath):
         os.remove(dskpath)  # pragma: no cover
     spice.kclear()
-    print("done dskw02 test")
 
 
 def test_dskx02():
@@ -2007,19 +2005,16 @@ def test_ekaced():
 
 
 def test_ekmany():
-    print("in ekmany")
     spice.kclear()
     ekpath = os.path.join(cwd, "example_ekmany.ek")
     tablename = "test_table_ekmany"
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     # Create new EK and new segment with table
-    print("call ekopn")
     handle = spice.ekopn(ekpath, ekpath, 0)
     decls = ["DATATYPE = CHARACTER*(10),   NULLS_OK = FALSE, SIZE = VARIABLE",
              "DATATYPE = DOUBLE PRECISION, NULLS_OK = FALSE, SIZE = VARIABLE",
              "DATATYPE = INTEGER,          NULLS_OK = FALSE, SIZE = VARIABLE"]
-    print("call ekbseg")
     segno = spice.ekbseg(handle, tablename, ['c1', 'd1', 'i1'], decls)
     # Insert records:  1, 2, and 3 entries at rows 0, 1, 2, respectively
     c_data = [['100'], ['101', '101'], ['102', '102', '102']]
@@ -2032,27 +2027,23 @@ def test_ekmany():
         spice.ekacei(handle, segno, r, "i1", len(i_data[r]), i_data[r], False)
     # Try record insertion beyond the next available, verify the exception
     with pytest.raises(spice.stypes.SpiceyError):
-        print("call ekinsr")
         spice.ekinsr(handle, segno, 4)
     # Close EK, then reopen for reading
-    print("call ekcls")
     spice.ekcls(handle)
     spice.kclear()
     #
     # Start of part two
     #
-    print("call eklef")
     handle = spice.eklef(ekpath)
     assert handle is not None
     # Test query using ekpsel
     query = "SELECT c1, d1, i1 from {}".format(tablename)
-    print("call ekpsel")
     n, xbegs, xends, xtypes, xclass, tabs, cols, err, errmsg = spice.ekpsel(query, 99, 99, 99)
     assert n == 3
-    assert spice.stypes.SpiceEKDataType._SPICE_CHR.value == xtypes[0]
-    assert spice.stypes.SpiceEKDataType._SPICE_DP.value  == xtypes[1]
-    assert spice.stypes.SpiceEKDataType._SPICE_INT.value == xtypes[2]
-    assert ([dict(spice.stypes.SpiceEKExprClass._fields_)['SPICE_EK_EXP_COL'].value]*3) == list(xclass)
+    assert spice.stypes.SpiceEKDataType.SPICE_CHR == xtypes[0]
+    assert spice.stypes.SpiceEKDataType.SPICE_DP  == xtypes[1]
+    assert spice.stypes.SpiceEKDataType.SPICE_INT == xtypes[2]
+    assert ([spice.stypes.SpiceEKExprClass.SPICE_EK_EXP_COL]*3) == list(xclass)
     assert (["TEST_TABLE_EKMANY"]*3) == tabs
     assert "C1 D1 I1".split() == cols
     assert not err
@@ -2108,7 +2099,6 @@ def test_ekmany():
         spice.ekrced(handle, segno, 3, "d1")
     #with pytest.raises(spice.stypes.SpiceyError): TODO: FIX
     #    spice.ekrcec(handle, segno, 4, "c1", 4) # this causes a SIGSEGV
-
     #
     # Part 3
     #
@@ -2152,7 +2142,6 @@ def test_ekmany():
     assert not spice.failed()
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
-    print("finished ekmany")
 
 
 def test_ekaclc():

--- a/spiceypy/tests/test_wrapper.py
+++ b/spiceypy/tests/test_wrapper.py
@@ -2016,40 +2016,44 @@ def test_ekmany():
     if spice.exists(ekpath):
         os.remove(ekpath) # pragma: no cover
     # Create new EK and new segment with table
+    print("call ekopn")
     handle = spice.ekopn(ekpath, ekpath, 0)
-    segno = spice.ekbseg(handle, tablename, 3, 10, "c1 d1 i1".split(), 200,
-                         ["DATATYPE = CHARACTER*(10),   NULLS_OK = FALSE, SIZE = VARIABLE",
-                          "DATATYPE = DOUBLE PRECISION, NULLS_OK = FALSE, SIZE = VARIABLE",
-                          "DATATYPE = INTEGER,          NULLS_OK = FALSE, SIZE = VARIABLE"]
-                        )
-    # Use EKINSR to insert two records at recno 0, one at recno 2
-    # - First inserted record will become final record 1
-    for insertrecno, finalrecno in ((0, 1), (0, 0), (2, 2)):
-        assert None is spice.ekinsr(handle, segno, insertrecno)
-        assert not spice.failed()
-        # Insert records:  1, 2, and 3 entries at rows 0, 1, 2, respectively
-        # - Integer values are final record number + 100
-        iBaseVals = [100+finalrecno]*(finalrecno+1)
-        # List-ify map to ensure listToCharArrayPtr sees list, not map object
-        spice.ekacec(handle, segno, insertrecno, "c1", finalrecno+1, 11, list(map(str, iBaseVals)),   False)
-        spice.ekaced(handle, segno, insertrecno, "d1", finalrecno+1,     list(map(float, iBaseVals)), False)
-        spice.ekacei(handle, segno, insertrecno, "i1", finalrecno+1,     iBaseVals,                   False)
-        assert not spice.failed()
+    decls = ["DATATYPE = CHARACTER*(10),   NULLS_OK = FALSE, SIZE = VARIABLE",
+             "DATATYPE = DOUBLE PRECISION, NULLS_OK = FALSE, SIZE = VARIABLE",
+             "DATATYPE = INTEGER,          NULLS_OK = FALSE, SIZE = VARIABLE"]
+    print("call ekbseg")
+    segno = spice.ekbseg(handle, tablename, ['c1', 'd1', 'i1'], decls)
+    # Insert records:  1, 2, and 3 entries at rows 0, 1, 2, respectively
+    c_data = [['100'], ['101', '101'], ['102', '102', '102']]
+    d_data = [[100.0], [101.0, 101.0], [102.0, 102.0, 102.0]]
+    i_data = [[100], [101, 101], [102, 102, 102]]
+    for r in range(0, 3):
+        spice.ekinsr(handle, segno, r)
+        spice.ekacec(handle, segno, r, "c1", len(c_data[r]), c_data[r], False)
+        spice.ekaced(handle, segno, r, "d1", len(d_data[r]), d_data[r], False)
+        spice.ekacei(handle, segno, r, "i1", len(i_data[r]), i_data[r], False)
     # Try record insertion beyond the next available, verify the exception
     with pytest.raises(spice.stypes.SpiceyError):
+        print("call ekinsr")
         spice.ekinsr(handle, segno, 4)
     # Close EK, then reopen for reading
+    print("call ekcls")
     spice.ekcls(handle)
     spice.kclear()
+    #
+    # Start of part two
+    #
+    print("call eklef")
     handle = spice.eklef(ekpath)
     assert handle is not None
     # Test query using ekpsel
     query = "SELECT c1, d1, i1 from {}".format(tablename)
+    print("call ekpsel")
     n, xbegs, xends, xtypes, xclass, tabs, cols, err, errmsg = spice.ekpsel(query, 99, 99, 99)
-    assert 3 == n
-    assert dict(spice.stypes.SpiceEKDataType._fields_)['SPICE_CHR'].value == xtypes[0]
-    assert dict(spice.stypes.SpiceEKDataType._fields_)['SPICE_DP'].value == xtypes[1]
-    assert dict(spice.stypes.SpiceEKDataType._fields_)['SPICE_INT'].value == xtypes[2]
+    assert n == 3
+    assert spice.stypes.SpiceEKDataType._SPICE_CHR.value == xtypes[0]
+    assert spice.stypes.SpiceEKDataType._SPICE_DP.value  == xtypes[1]
+    assert spice.stypes.SpiceEKDataType._SPICE_INT.value == xtypes[2]
     assert ([dict(spice.stypes.SpiceEKExprClass._fields_)['SPICE_EK_EXP_COL'].value]*3) == list(xclass)
     assert (["TEST_TABLE_EKMANY"]*3) == tabs
     assert "C1 D1 I1".split() == cols
@@ -2060,102 +2064,91 @@ def test_ekmany():
     assert nmrows == 3
     assert not error
     assert '' == errmsg
+    # test fail case for eknelt
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.eknelt(0, nmrows+1)
     # Validate the content of each field, including exceptions when
-    saveCData = []
-    saveDData = []
-    saveIData = []
     # Loop over rows, test .ekgc/.ekgd/.ekgi
-    for recno in range(nmrows+1):
-        # Confirm number of elements in each row
-        if recno == nmrows:
-            with pytest.raises(spice.stypes.SpiceyError):
-                spice.eknelt(0, recno)
-        else:
-            assert (recno+1) == spice.eknelt(0, recno)
-        # Confirm the elements' values
-        saveCData.append([])
-        saveDData.append([])
-        saveIData.append([])
-        for iElement in range(recno+2):
-            if iElement > recno or recno == nmrows:
-                with pytest.raises(spice.stypes.SpiceyError):
-                    spice.ekgc(0, recno, iElement, lenout=11)
-                if spice.failed(): spice.reset()
-                with pytest.raises(spice.stypes.SpiceyError):
-                    spice.ekgd(1, recno, iElement)
-                if spice.failed(): spice.reset()
-                with pytest.raises(spice.stypes.SpiceyError):
-                    spice.ekgi(2, recno, iElement)
-                if spice.failed(): spice.reset()
-            else:
-                cData, iNull = spice.ekgc(0, recno, iElement, lenout=11)
-                assert cData == str(100+recno)
-                assert iNull == False
-                saveCData[-1].append(cData)
-                dData, iNull = spice.ekgd(1, recno, iElement)
-                assert dData == (100.+recno)
-                assert iNull == False
-                saveDData[-1].append(dData)
-                iData, iNull = spice.ekgi(2, recno, iElement)
-                assert iData == (100+recno)
-                assert iNull == False
-                saveIData[-1].append(iData)
+    for r in range(nmrows):
+        # get number of elements in this row
+        n_elm = spice.eknelt(0, r)
+        assert  n_elm == r + 1
+        for e in range(0, n_elm):
+            # get row int data
+            i_datum, i_null = spice.ekgi(2, r, e)
+            assert not i_null
+            assert i_datum == i_data[r][e]
+            # get row double data
+            d_datum, d_null = spice.ekgd(1, r, e)
+            assert not d_null
+            assert d_datum == d_data[r][e]
+            # get row char data
+            c_datum, c_null = spice.ekgc(0, r, e)
+            assert not c_null
+            assert c_datum == c_data[r][e]
     # Loop over rows, test .ekrcec/.ekrced/.ekrcei
-    for recno in range(nmrows+1):
-        if recno == nmrows:
-            #with pytest.raises(spice.stypes.SpiceyError):
-            #    spice.ekrcec(handle, segno, recno, "c1", 20, nelts=2222)
-            #if spice.failed(): spice.reset()
-            with pytest.raises(spice.stypes.SpiceyError):
-                spice.ekrced(handle, segno, recno, "d1")
-            if spice.failed(): spice.reset()
-            with pytest.raises(spice.stypes.SpiceyError):
-                spice.ekrcei(handle, segno, recno, "i1")
-            if spice.failed(): spice.reset()
-        else:
-            nvals, altCData, isNull = spice.ekrcec(handle, segno, recno, "c1", 11, nelts=len(saveCData[recno]))
-            assert (recno+1) == nvals
-            assert list(altCData) == saveCData[recno]
-            nvals, altDData, isNull = spice.ekrced(handle, segno, recno, "d1", nelts=len(saveDData[recno]))
-            assert (recno+1) == nvals
-            assert list(altDData) == saveDData[recno]
-            nvals, altIData, isNull = spice.ekrcei(handle, segno, recno, "i1", nelts=len(saveIData[recno]))
-            assert (recno+1) == nvals
-            assert list(altIData) == saveIData[recno]
+    for r in range(nmrows):
+        # get row int data
+        ni_vals, ri_data, i_null = spice.ekrcei(handle, segno, r, "i1")
+        assert not i_null
+        assert ni_vals == r + 1
+        npt.assert_array_equal(ri_data, i_data[r])
+        # get row double data
+        nd_vals, rd_data, d_null = spice.ekrced(handle, segno, r, "d1")
+        assert not d_null
+        assert nd_vals == r + 1
+        npt.assert_array_equal(rd_data, d_data[r])
+        # get row char data
+        nc_vals, rc_data, c_null = spice.ekrcec(handle, segno, r, "c1", 11)
+        assert not c_null
+        assert nc_vals == r + 1
+        assert rc_data == c_data[r]
+    # test out of bounds
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.ekrcei(handle, segno, 3, "i1")
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.ekrced(handle, segno, 3, "d1")
+    #with pytest.raises(spice.stypes.SpiceyError): TODO: FIX
+    #    spice.ekrcec(handle, segno, 4, "c1", 4) # this causes a SIGSEGV
+
+    #
+    # Part 3
+    #
     # Close file, re-open for writing
     spice.ekuef(handle)
     handle = spice.ekopw(ekpath)
     # Loop over rows, update values using .ekucec/.ekuced/.ekucei
-    for recno in range(nmrows+1):
-        iBaseVals = [200+recno]*(recno+1)
-        if recno == nmrows:
-            #with pytest.raises(spice.stypes.SpiceyError):
-            #    spice.ekucec(handle, segno, recno, "c1", recno+1, 11, list(map(str, iBaseVals)), False)
-            #if spice.failed(): spice.reset()
-            with pytest.raises(spice.stypes.SpiceyError):
-                spice.ekuced(handle, segno, recno, "d1", recno+1, list(map(float, iBaseVals)), False)
-            if spice.failed(): spice.reset()
-            with pytest.raises(spice.stypes.SpiceyError):
-                spice.ekucei(handle, segno, recno, "i1", recno+1, iBaseVals, False)
-            if spice.failed(): spice.reset()
-        else:
-            # List-ify map to ensure listToCharArrayPtr sees list, not map object
-            spice.ekucec(handle, segno, recno, "c1", recno+1, 11, list(map(str, iBaseVals)), False)
-            spice.ekuced(handle, segno, recno, "d1", recno+1, list(map(float, iBaseVals)), False)
-            spice.ekucei(handle, segno, recno, "i1", recno+1, iBaseVals, False)
-            assert not spice.failed()
+    c_data = [['200'], ['201', '201'], ['202', '202', '202']]
+    d_data = [[200.0], [201.0, 201.0], [202.0, 202.0, 202.0]]
+    i_data = [[200], [201, 201], [202, 202, 202]]
+    for r in range(0, 3):
+        spice.ekucec(handle, segno, r, "c1", len(c_data[r]), c_data[r], False)
+        spice.ekuced(handle, segno, r, "d1", len(d_data[r]), d_data[r], False)
+        spice.ekucei(handle, segno, r, "i1", len(i_data[r]), i_data[r], False)
+    # Test invalid updates
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.ekucec(handle, segno, 3, "c1", 1, ['300'], False)
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.ekuced(handle, segno, 3, "d1", 1, [300.0], False)
+    with pytest.raises(spice.stypes.SpiceyError):
+        spice.ekucei(handle, segno, 3, "i1", 1, [300], False)
     # Loop over rows, use .ekrcec/.ekrced/.ekrcei to test updates
-    for recno in range(nmrows):
-        iBaseVals = [200+recno]*(recno+1)
-        nvals, altCData, isNull = spice.ekrcec(handle, segno, recno, "c1", 11, nelts=len(saveCData[recno]))
-        assert (recno+1) == nvals
-        assert list(altCData) == list(map(str, iBaseVals))
-        nvals, altDData, isNull = spice.ekrced(handle, segno, recno, "d1", nelts=len(saveDData[recno]))
-        assert (recno+1) == nvals
-        assert list(altDData) == iBaseVals
-        nvals, altIData, isNull = spice.ekrcei(handle, segno, recno, "i1", nelts=len(saveIData[recno]))
-        assert (recno+1) == nvals
-        assert list(altIData) == list(map(float, iBaseVals))
+    for r in range(nmrows):
+        # get row int data
+        ni_vals, ri_data, i_null = spice.ekrcei(handle, segno, r, "i1")
+        assert not i_null
+        assert ni_vals == r + 1
+        npt.assert_array_equal(ri_data, i_data[r])
+        # get row double data
+        nd_vals, rd_data, d_null = spice.ekrced(handle, segno, r, "d1")
+        assert not d_null
+        assert nd_vals == r + 1
+        npt.assert_array_equal(rd_data, d_data[r])
+        # get row char data
+        nc_vals, rc_data, c_null = spice.ekrcec(handle, segno, r, "c1", 11)
+        assert not c_null
+        assert nc_vals == r + 1
+        assert rc_data == c_data[r]
     # Cleanup
     spice.ekcls(handle)
     assert not spice.failed()

--- a/spiceypy/utils/support_types.py
+++ b/spiceypy/utils/support_types.py
@@ -109,7 +109,7 @@ def toPythonString(inString):
     elif six.PY3:
         if isinstance(inString, c_char_p):
             return toPythonString(inString.value)
-        return bytes.decode(string_at(inString), encoding='utf-8', errors="ignore").rstrip()
+        return bytes.decode(string_at(inString), errors="ignore").rstrip()
 
 
 def emptyCharArray(xLen=None, yLen=None):

--- a/spiceypy/utils/support_types.py
+++ b/spiceypy/utils/support_types.py
@@ -592,6 +592,7 @@ class SpiceEKDataType(c_int):
     _SPICE_TIME = c_int(3)
     _SPICE_BOOL = c_int(4)
 
+
     _fields_ = [
         ('SPICE_CHR', _SPICE_CHR),
         ('SPICE_DP', _SPICE_DP),
@@ -599,6 +600,12 @@ class SpiceEKDataType(c_int):
         ('SPICE_TIME', _SPICE_TIME),
         ('SPICE_BOOL', _SPICE_BOOL),
     ]
+
+    SPICE_CHR =  _SPICE_CHR.value
+    SPICE_DP  =  _SPICE_DP.value
+    SPICE_INT =  _SPICE_INT.value
+    SPICE_TIME = _SPICE_TIME.value
+    SPICE_BOOL = _SPICE_BOOL.value
 
 
 def emptySpiceEKDataTypeVector(n):
@@ -609,11 +616,19 @@ def emptySpiceEKDataTypeVector(n):
 
 
 class SpiceEKExprClass(c_int):
+    _SPICE_EK_EXP_COL = c_int(0)
+    _SPICE_EK_EXP_FUNC = c_int(1)
+    _SPICE_EK_EXP_EXPR = c_int(2)
+
     _fields_ = [
-        ('SPICE_EK_EXP_COL', c_int(0)),
-        ('SPICE_EK_EXP_FUNC', c_int(1)),
-        ('SPICE_EK_EXP_EXPR', c_int(2))
+        ('SPICE_EK_EXP_COL',  _SPICE_EK_EXP_COL),
+        ('SPICE_EK_EXP_FUNC', _SPICE_EK_EXP_FUNC),
+        ('SPICE_EK_EXP_EXPR', _SPICE_EK_EXP_EXPR)
     ]
+
+    SPICE_EK_EXP_COL = _SPICE_EK_EXP_COL.value
+    SPICE_EK_EXP_FUNC = _SPICE_EK_EXP_FUNC.value
+    SPICE_EK_EXP_EXPR = _SPICE_EK_EXP_EXPR.value
 
 
 def emptySpiceEKExprClassVector(n):

--- a/spiceypy/utils/support_types.py
+++ b/spiceypy/utils/support_types.py
@@ -109,7 +109,7 @@ def toPythonString(inString):
     elif six.PY3:
         if isinstance(inString, c_char_p):
             return toPythonString(inString.value)
-        return bytes.decode(string_at(inString)).rstrip()
+        return bytes.decode(string_at(inString), encoding='utf-8', errors="ignore").rstrip()
 
 
 def emptyCharArray(xLen=None, yLen=None):
@@ -586,12 +586,18 @@ class SpiceDLADescr(Structure):
 
 
 class SpiceEKDataType(c_int):
+    _SPICE_CHR = c_int(0)
+    _SPICE_DP  = c_int(1)
+    _SPICE_INT = c_int(2)
+    _SPICE_TIME = c_int(3)
+    _SPICE_BOOL = c_int(4)
+
     _fields_ = [
-        ('SPICE_CHR', c_int(0)),
-        ('SPICE_DP', c_int(1)),
-        ('SPICE_INT', c_int(2)),
-        ('SPICE_TIME', c_int(3)),
-        ('SPICE_BOOL', c_int(4)),
+        ('SPICE_CHR', _SPICE_CHR),
+        ('SPICE_DP', _SPICE_DP),
+        ('SPICE_INT', _SPICE_INT),
+        ('SPICE_TIME', _SPICE_TIME),
+        ('SPICE_BOOL', _SPICE_BOOL),
     ]
 
 

